### DISCUSSION TITLE
mail作业task7failed的一个解决方案（原理不明）

### DISCUSSION THREAD

**Discussion by Exiurs** [physics-data/faq#130] [Ideas] [answered]
**大家别用这个方法，感觉能过只是巧合，还是想一下怎么不改target吧**
[https://physics-data.meow.plus/faq/error/mail/](url)

目前本方案已帮助三名同学
在Makefile的代码块中，使用`$(shell ./*.sh)`执行外部脚本，而不要直接使用`./*.sh`或是`bash ./*.sh`
同时在每个脚本开头加上`#!/bin/bash`
记得在make或是测试grade或是git前在`./mail-xxx/`目录下使用`chmod +x *.sh`给每个脚本添加权限（否则可能出现permission denied报错）
目前不知道原理是啥，但确实在使用后task7从failed到passed
（只记得上课好像讲过调用外部脚本要用shell函数
（贴一下网上找到的教程
![微信截图_20210905020306](https://user-images.githubusercontent.com/85775244/132104184-6fd8e740-fa5a-41ba-bcdd-69ef7a11a2c8.png)
[https://seisman.github.io/how-to-write-makefile/functions.html#id4](url)
（欢迎大家讨论

**Comment by heroxbd** [comment]
脚本的可执行权限原定上课要讲，但是没有来得及。

一个脚本具有的执行权限需要额外赋予，这是出于安全的考虑，防止我们在不知情时误执行不明代码。因此在脚本被标记为 `+x` 之前，会有 permission denied 的报错。

浏览了一遍 https://seisman.github.io/how-to-write-makefile ，我觉得 Make 的官方文档更值得推荐， `info make` 可达。

**Comment by YiHui-Liu** [comment, marked as answer]
文档链接：[shell函数](https://seisman.github.io/how-to-write-makefile/functions.html#shell)
<hr />

关于task7，还可能存在`* is up to date`，`cat *.sh>* chmod a+x *.sh`等问题，解决方法可参考官方文档：[Phony Targets](https://www.gnu.org/software/make/manual/html_node/Phony-Targets.html)，可能的原因是因为：
>Your have directories with the same name as your targets.
Since your targets don't depend on anything, they'll always be considered up to date.

GNU make默认每一个target都是文件，在执行`make target`指令时，会优先检测当前目录下是否存在该文件（在本例中，同名的则是`important`和`reply`）。如果存在，make就不会执行commands，否则就会出现上述情况。

**Reply by heroxbd** [reply]
可以参考 [Phony Targets](https://www.gnu.org/software/make/manual/html_node/Phony-Targets.html)

**Reply by phy-void** [reply]
分享一个task7的解决方案，在@YiHui-Liu帮助下解决：
关于一种解决重复make的思路是判断make之后产生的文件是否存在，若存在则不执行；这里如果使用shell的命令进行判断，运行make时均可正确执行功能，但make -q并不会通过。正确的解决方法是使用makefile的内建条件判断来完成。希望对遇到类似问题的同学有所帮助。对于make -q的原理还没有完全理解。

相关语法参考：
https://seisman.github.io/how-to-write-makefile/conditionals.html
https://blog.csdn.net/wwwlyj123321/article/details/107665968
